sinonjs/samsam
# Distribute package as source

@khamiltonuk and I spent some time trying to replace `rollup` with `esbuild` only to [discover that it drops comments](https://github.com/evanw/esbuild/issues/578). That would bundling with `esbuild` and providing `.d.ts` files mutually exclusive.

I realised that we don't actually need bundling at all in this project.

---

This library is not meant for writing end user applications or even for being used directly when writing tests. It is not meant to be loaded directly by browsers.

Consumers of this package are assumed to use their own bundlers, should they need to bundle code for use in browsers or workers.

Therefore, it is safe to distribute the package as source files and not bundle them up.

This allows us to remove the bundler and its transitive dependencies, which reduces the maintenance burden of managing the library.

#### Background

* We'd like to ship the library with `.d.ts` files, to allow improved experience for TypeScript users
* We would like to reduce maintenance burden of the library

#### Solution

* Distribute the package as source files
* Create `.d.ts` files from the source files
* Remove the bundler

#### How to verify

1. Check out this branch
1. `npm ci`
1. `npm run build`
1. Observe that there are `.d.ts` files in the `types/` folder
2. `npm pack`
3. Observe that the `.d.ts` files are included in the package
4. `npm link`
5. In a `sinon` checkout:
  1.  `npm link @sinonjs/samsam`
  2. `npm test`
  3. Observe that tests still pass
